src/Http/Http/src/Timeouts/RequestTimeoutsMiddleware.cs

```diff
@@ -0,0 +1,132 @@
+        {
+            return _next(context);
+        }
+
+        var disableMetadata = endpoint?.Metadata.GetMetadata<DisableRequestTimeoutAttribute>();
+        if (disableMetadata is not null)
+        {
+            return _next(context);
+        }
+
+        RequestTimeoutPolicy? selectedPolicy = null;
+        TimeSpan? timeSpan = null;
+        if (policyMetadata is not null)
+        {
+            selectedPolicy = policyMetadata;
+        }
+        else if (timeoutMetadata is not null)
+        {
+            if (timeoutMetadata.Timeout is not null)
+            {
+                timeSpan = timeoutMetadata.Timeout.Value;
+            }
+            else
+            {
+                if (options.Policies.TryGetValue(timeoutMetadata.PolicyName!, out var policy))
+                {
+                    selectedPolicy = policy;
+                }
+                else
+                {
+                    throw new InvalidOperationException($"The requested timeout policy '{timeoutMetadata.PolicyName}' is not available.");
+                }
+            }
+        }
+        else
+        {
+            selectedPolicy = options.DefaultPolicy;
+        }
+
+        if (timeSpan is null && selectedPolicy?.Timeout is not null)
+        {
+            timeSpan = selectedPolicy.Timeout;
+        }
+
+        if (timeSpan is null || timeSpan == Timeout.InfiniteTimeSpan)
+        {
+            return _next(context);
+        }
+
+        return SetTimeoutAsync();
+
+        async Task SetTimeoutAsync()
+        {
+            var originalToken = context.RequestAborted;
+            using var linkedCts = _cancellationTokenProvider.GetLinkedCancellatinTokeSource(context, originalToken, timeSpan.Value);
+            try
+            {
+                context.RequestAborted = linkedCts.Token;
+                await _next(context);
+            }
+            catch (OperationCanceledException operationCanceledException)
+            {
+                if (!context.Response.HasStarted &&
+                    linkedCts.Token.IsCancellationRequested &&
+                    !originalToken.IsCancellationRequested)
+                {
+                    _logger.TimeoutExceptionHandled(operationCanceledException);
+
+                    context.Response.Clear();
+
+                    if (selectedPolicy?.TimeoutStatusCode is null)
+                    {
+                        context.Response.StatusCode = StatusCodes.Status504GatewayTimeout;
+                    }
+                    else
+                    {
+                        context.Response.StatusCode = selectedPolicy.TimeoutStatusCode.Value;
+                    }
+
+                    if (selectedPolicy?.WriteTimeoutResponse is not null)
+                    {
+                        await selectedPolicy.WriteTimeoutResponse(context);
+                    }
+                }
+                else
+                {
+                    throw;
+                }
```

Comment:
Inverting to flatten
```suggestion
                if (context.Response.HasStarted ||
                    !linkedCts.Token.IsCancellationRequested ||
                    originalToken.IsCancellationRequested)
                {
                    // We can't produce a response, or it wasn't our timeout that caused this.
                    throw;
                }
                
                _logger.TimeoutExceptionHandled(operationCanceledException);

                context.Response.Clear();

                context.Response.StatusCode = selectedPolicy?.TimeoutStatusCode ?? StatusCodes.Status504GatewayTimeout;

                if (selectedPolicy?.WriteTimeoutResponse is not null)
                {
                    await selectedPolicy.WriteTimeoutResponse(context);
                }
```